### PR TITLE
WIP: Feature/23 serializable actions

### PR DESCRIPTION
**Pull request by mweststrate**

### CODE REVIEW

File: src/types/reference.ts [hunk 1]
```diff
@@ -16,7 +16,10 @@ export interface IReferenceDescription {
 
 export function referenceTo<T>(path: string): T;
 export function referenceTo<T>(getter: IReferenceGetter<T>, setter?: IReferenceSetter<T>): T;
+export function referenceTo<T>(factory: IFactory<any, T>): T;
 export function referenceTo(arg1, arg2?) {
```
Comment (mattiamanzati) [inline src/types/reference.ts]: TODO: Maybe export as "reference" to be consistent with other types and more readable?

File: src/types/reference.ts [hunk 2]
```diff
@@ -27,17 +30,38 @@ export function referenceTo(arg1, arg2?) {
...
+        },
+        (value: any, owner: IModel, name) => {
+            if (value === null || value === undefined)
+                return value
+            invariant(isModel(value), `The value assigned to the reference '${name}' is not a model instance`)
+            invariant(factory.is(value), `The value assigned to the reference '${name}' is not a model of type ${factory}`)
+            const base = getNode(owner)
+            const target = getNode(value)
+            invariant(base.root === target.root, `The value assigned to the reference '${name}' should already be part of the same model tree`)
+            return getRelativePath(base, target)
```
Comment (mattiamanzati) [inline src/types/reference.ts]: TODO: Maybe return as {$path: "..."} or {$ref: "..."}

File: src/core/action.ts [hunk 3]
```diff
@@ -22,58 +16,80 @@ export function createNonActionWrapper(instance, key, func) {
...
+        return arg
+    } catch (e) {
+        throw new Error(`Argument ${index} that was passed to action '${actionName}' is not serializable.`)
+    }
+}
+
+function deserializeArgument(adm: Node, value: any): any {
+    if (typeof value === "object") {
+        const keys = Object.keys(value)
+        if (keys.length === 1 && keys[0] === "$path")
```
Comment (mweststrate) [inline src/core/action.ts]: done

File: src/types/reference.ts (hunk 1 above)
Comment (mweststrate) [inline src/types/reference.ts]: done

File: src/types/reference.ts (hunk 2 above)
Comment (mweststrate) [inline src/types/reference.ts]: Added TODO to code

**Pull request merged**